src/cata_tiles.cpp

```diff
@@ -1572,27 +1573,67 @@ void cata_tiles::draw( const point &dest, const tripoint &center, int width, int
+                &cata_tiles::draw_field_or_item,
+                &cata_tiles::draw_vpart_no_roof, &cata_tiles::draw_vpart_roof,
+                &cata_tiles::draw_critter_at, &cata_tiles::draw_zone_mark,
+                &cata_tiles::draw_zombie_revival_indicators
+            }
+        };
+        // Legacy code for isometric tilesets until they are ready
+        const std::array<decltype( &cata_tiles::draw_furniture ), 14> drawing_layers_legacy = {{
+                &cata_tiles::draw_terrain, &cata_tiles::draw_furniture, &cata_tiles::draw_graffiti, &cata_tiles::draw_trap, &cata_tiles::draw_part_con,
                 &cata_tiles::draw_field_or_item, &cata_tiles::draw_vpart_below,
                 &cata_tiles::draw_critter_at_below, &cata_tiles::draw_terrain_below,
                 &cata_tiles::draw_vpart_no_roof, &cata_tiles::draw_vpart_roof,
                 &cata_tiles::draw_critter_at, &cata_tiles::draw_zone_mark,
                 &cata_tiles::draw_zombie_revival_indicators
             }
         };
-        // for each of the drawing layers in order, back to front ...
-        for( auto f : drawing_layers ) {
-            // ... draw all the points we drew terrain for, in the same order
+        if( is_isometric() ) {
+            // Disable multi z-level display on isometric tilesets. Can be enabled but will cause the following issues:
+            // 1. Supposedly transparent terrain and fields such as t_open_air need to be given sprites
+            // Otherwise they will be replaced with an opaque fallback sprite
+            // 2. May worsen visibility issues in isometric tilesets with inconsistent terrain heights
             for( tile_render_info &p : draw_points ) {
-                ( this->*f )( p.pos, p.ll, p.height_3d, p.invisible );
+                for( auto f : drawing_layers_legacy ) {
+                    ( this->*f )( p.pos, p.ll, p.height_3d, p.invisible );
+                }
+            }
+        } else {
+            // For every point we drew terrain for ...
+            for( tile_render_info &p : draw_points ) {
+                // Find lowest z-level to draw
+                tripoint p_draw = p.pos;
+                int cur_height_3d = p.height_3d;
+                while( !here.dont_draw_lower_floor( p_draw ) && p_draw.z > -10 ) {
+                    p_draw.z -= 1;
+                    cur_height_3d -= 1;
+                }
+
+                // Draw all layers for the bottom z-level
+                for( auto f : drawing_layers ) {
+                    ( this->*f )( p_draw, p.ll, cur_height_3d, p.invisible );
+                }
+
+                // Keep going up and drawing until current position
+                while( p_draw.z < p.pos.z ) {
+                    // Draws fog overlay to give the impression of distance
+                    draw_zlevel_overlay( p_draw, p.ll, color_blocks );
```

Comment:
You can compute the alpha and draw the overlay only once.
I used `int alpha = 77.3059 * std::log(3.9131 * std::abs( p.pos.z - lowest ));` as an approximation but you can also compute the real blended alpha with `dstA = srcA + (dstA * (1-srcA))` in a loop.
<details>
<summary>before, with multiple rects - staring down from z-level 6</summary>

![6b](https://github.com/CleverRaven/Cataclysm-DDA/assets/68240139/28e03cd5-2a80-409d-8907-7b4f214604b3)


</details>

<details>
<summary>after, with alpha approximation</summary>

![6a](https://github.com/CleverRaven/Cataclysm-DDA/assets/68240139/66eeaa23-10b2-4955-be01-9282b071d0e8)


</details>

<details>
<summary>patch</summary>

```diff
diff --git a/src/cata_tiles.cpp b/src/cata_tiles.cpp
index 2661c55e27..070c6105c7 100644
--- a/src/cata_tiles.cpp
+++ b/src/cata_tiles.cpp
@@ -1615,6 +1615,7 @@ void cata_tiles::draw( const point &dest, const tripoint &center, int width, int
                     p_draw.z -= 1;
                     cur_height_3d -= 1;
                 }
+                int const lowest = p_draw.z;
 
                 // Draw all layers for the bottom z-level
                 for( auto f : drawing_layers ) {
@@ -1623,14 +1624,18 @@ void cata_tiles::draw( const point &dest, const tripoint &center, int width, int
 
                 // Keep going up and drawing until current position
                 while( p_draw.z < p.pos.z ) {
-                    // Draws fog overlay to give the impression of distance
-                    draw_zlevel_overlay( p_draw, p.ll, color_blocks );
                     p_draw.z += 1;
                     cur_height_3d += 1;
                     for( auto f : drawing_layers ) {
                         ( this->*f )( p_draw, p.ll, cur_height_3d, p.invisible );
                     }
                 }
+                
+                if( lowest != p.pos.x ) {
+                    int alpha = 77.3059 * std::log(3.9131 * std::abs( p.pos.z - lowest ));
+                    // Draws fog overlay to give the impression of distance
+                    draw_zlevel_overlay( p_draw, p.ll, color_blocks, alpha );
+                }
             }
         }
 
@@ -3841,7 +3846,7 @@ bool cata_tiles::draw_zombie_revival_indicators( const tripoint &pos, const lit_
 }
 
 void cata_tiles::draw_zlevel_overlay( const tripoint &p, const lit_level ll,
-                                      color_block_overlay_container &color_blocks )
+                                      color_block_overlay_container &color_blocks, int alpha )
 {
     if( is_isometric() ) {
         return;
@@ -3862,7 +3867,7 @@ void cata_tiles::draw_zlevel_overlay( const tripoint &p, const lit_level ll,
             fog_color = curses_color_to_SDL( c_dark_gray );
         }
         // Setting for fog transparancy
-        fog_color.a = 100;
+        fog_color.a = alpha;
 
         // Transparancy will only work in blend mode
         color_blocks.first = SDL_BLENDMODE_BLEND;
diff --git a/src/cata_tiles.h b/src/cata_tiles.h
index 7b5093147d..5af01574dc 100644
--- a/src/cata_tiles.h
+++ b/src/cata_tiles.h
@@ -533,7 +533,7 @@ class cata_tiles
         bool draw_zombie_revival_indicators( const tripoint &pos, lit_level ll, int &height_3d,
                                              const std::array<bool, 5> &invisible );
         void draw_zlevel_overlay( const tripoint &p, lit_level ll,
-                                  color_block_overlay_container &color_blocks );
+                                  color_block_overlay_container &color_blocks, int alpha );
         void draw_entity_with_overlays( const Character &ch, const tripoint &p, lit_level ll,
                                         int &height_3d );
 
```

</details>

Also, the function is a little misnamed as it doesn't draw anything itself.

Reply:
I actually did something similar in my earlier iterations, but I wanted to make sure that the z-level of flying creatures are accurately represented.  Now that I've done some testing, I realized that color blocks tend to draw at the top-most layer regardless, which is something I'll have to look into, but at least for custom fog textures, drawing multiple layers will allow us to accurately perceive airborne objects.

![bee](https://github.com/CleverRaven/Cataclysm-DDA/assets/129854247/a71e7414-9b61-4ea6-8966-f632f8dc0e2f)

I get that this will do a bit more drawing though.  Think it is worth it?

Reply:
You could probably draw the overlay only once with alpha approximation when there's actually something else to draw on a particular z-level, so when there's only one creature in between, only two overlay draws are need. Not sure how to check whether there's something to redraw though given the current code calls the redraw functions unconditionally.

Reply:
I believe the draw functions are booleans which only return true if something is drawn, so perhaps it is possible.  Still, I'd like to find a way to accurately measure performance before I proceed with optimization.  Might experiment with the benchmark feature a bit.

Reply:
Tried to implement this but it didn't go well with custom sprites.  Forcing the alpha value of sprites to a predetermined value would simply cause them to display wrongly unless they were completely opaque in the first place.  My latest performance tests show that the majority of additional draw time is attributable to drawing the extra z-levels rather than the fog, so the performance improvement would have been limited anyways.